contrib/kythe/src/python/pants/contrib/kythe/tasks/extract_java.py

```diff
@@ -0,0 +1,117 @@
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+
+from pants.backend.jvm.subsystems.jvm import JVM
+from pants.backend.jvm.subsystems.shader import Shader
+from pants.backend.jvm.tasks.jvm_tool_task_mixin import JvmToolTaskMixin
+from pants.base.exceptions import TaskError
+from pants.java.jar.jar_dependency import JarDependency
+
+from pants.contrib.kythe.tasks.indexable_java_targets import IndexableJavaTargets
+
+
+# Kythe requires various system properties to be set (sigh).  So we can't use nailgun.
+class ExtractJava(JvmToolTaskMixin):
+  _KYTHE_EXTRACTOR_MAIN = 'com.google.devtools.kythe.extractors.java.standalone.Javac8Wrapper'
+
+  cache_target_dirs = True
+
+  @classmethod
+  def implementation_version(cls):
+    # Bump this version to invalidate all past artifacts generated by this task.
+    return super(ExtractJava, cls).implementation_version() + [('KytheExtract', 6), ]
+
+  @classmethod
+  def subsystem_dependencies(cls):
+    return super(ExtractJava, cls).subsystem_dependencies() + (JVM.scoped(cls),)
+
+  @classmethod
+  def product_types(cls):
+    # TODO: Support indexpack files?
+    return ['kindex_files']
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    super(ExtractJava, cls).prepare(options, round_manager)
+    round_manager.require_data('runtime_classpath')
+    round_manager.require_data('zinc_args')
+
+  @classmethod
+  def register_options(cls, register):
+    super(ExtractJava, cls).register_options(register)
+    cls.register_jvm_tool(register,
+                          'kythe-extractor',
+                          classpath=[JarDependency('kythe', 'extractors/javac_extractor',
+                                                   'v0.0.26-snowchain3')],
+                          custom_rules=[
+                            # These need to remain unshaded so that Kythe can interact with the
+                            # javac embedded in its jar.
+                            Shader.exclude_package('com.sun', recursive=True),
+                          ],
+                          main=cls._KYTHE_EXTRACTOR_MAIN)
+
+  def execute(self):
+    indexable_targets = IndexableJavaTargets.get(self.context)
+    targets_to_zinc_args = self.context.products.get_data('zinc_args')
+
+    with self.invalidated(indexable_targets, invalidate_dependents=True) as invalidation_check:
+      cp = self.tool_classpath('kythe-extractor')
+      for vt in invalidation_check.invalid_vts:
+        self.context.log.info('Kythe extracting from {}\n'.format(vt.target.address.spec))
+        javac_args = self._get_javac_args_from_zinc_args(targets_to_zinc_args[vt.target])
+        jvm_options = list(JVM.scoped_instance(self).get_jvm_options())
+        jvm_options.extend([
+          '-DKYTHE_CORPUS={}'.format(vt.target.address.spec),
```

Comment:
Funny, I recently changed these from env vars to sysprops, thinking that nailgun would be able to set the latter per-call. But I don't actually know what NG can do either way. Will postpone for a future change, once we figure out what NG supports?